src/protocol/input.rs

```diff
@@ -0,0 +1,340 @@
+use crate::bits::{BitArray, BitArray40, BitArray8};
+use crate::error::Error;
+use crate::ff::Field;
+use crate::helpers::Role;
+use crate::protocol::sort::apply_sort::shuffle::Resharable;
+use crate::protocol::Substep;
+use crate::protocol::{context::Context, RecordId};
+use crate::secret_sharing::replicated::semi_honest::{
+    AdditiveShare as Replicated, XorShare as XorReplicated,
+};
+use async_trait::async_trait;
+use futures::future::{try_join4, try_join_all};
+use futures::FutureExt;
+use std::iter::{repeat, zip};
+
+pub type MatchKey = BitArray40;
+pub type BreakdownKey = BitArray8;
+
+// Struct that holds all possible fields of the input to IPA.
+#[derive(Debug)]
+pub struct GenericReportShare<F: Field, MK: BitArray, BK: BitArray> {
```

Comment:
I think that this sort of catch-all type isn't the best way to approach this.

Taking a high-level look at this, we have a protocol that glues a bunch of discrete stages together into a coherent whole.  That overall protocol will have an input of a particular form (this with some fields removed and other fields made non-optional).

Internally, we need something as well.  Internal steps in that process carry stuff along to future steps, but don't care what that payload looks like (we can use a very generic `T: Resharable` or something for the stuff that we sort).  And some protocol parts directly chain together with concrete types.

Having a single type with lots of optional bits makes it hard to understand because the type doesn't force values to be valid in any given context.  Can I provide something with only breakdown keys and trigger values?  This type permits that.

It would be better overall to have each internal step define its own input format, then to try to rationalize them throughout by making them consistent with each other (with naming, types, etc...).

Reply:
Reducing the number of `*InputRow`/`*OutputRow` was something @benjaminsavage and I discussed, but I do agree that there's a benefit of having distinct structs for each protocol.

My motivation was to reduce duplicate trait implementations like `IntoShares`, `Reconstruct` and `Resharable`. Since two of those three are for tests, I've changed my approach a bit. We still have the generic structs but only for tests, while keeping the distinct input/output structs for production code. I think the protocols' data flow are easier to understand with this new change, while reducing some of the redundancy.